File: API-Biblioteca-DevEficiente-CDD/src/main/java/io/github/monthalcantara/bibliotecaCDD/livro/NovoLivroRequest.java
package io.github.monthalcantara.bibliotecaCDD.livro;

import com.fasterxml.jackson.annotation.JsonProperty;
import io.github.monthalcantara.bibliotecaCDD.utils.validation.ValorUnico;

import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import java.math.BigDecimal;

public class NovoLivroRequest {

    @NotBlank
    @JsonProperty("titulo")
    private String titulo;

    @NotNull
    @JsonProperty("preco")
    private BigDecimal preco;

    @NotBlank
    @ValorUnico(domainClass = Livro.class, fieldName = "isbn", message = "Ja existe um Livro com esse Isbn cadastrado")
    @JsonProperty("isbn")
    private String isbn;

    @Deprecated
    private NovoLivroRequest() {
    }

    public NovoLivroRequest(@NotBlank String titulo, @NotNull BigDecimal preco, @NotBlank String isbn) {
        this.titulo = titulo;
        this.preco = preco;
        this.isbn = isbn;
    }

    public Livro toModel() {
        return new Livro(this.titulo, this.preco, this.isbn);
    }

    public String getTitulo() {
        return titulo;
    }

    public BigDecimal getPreco() {
        return preco;
    }

    public String getIsbn() {
        return isbn;
    }
}

File: Kotlin/API-Casa_Do_Codigo_Kotlin/src/main/kotlin/io/github/monthalcantara/casadocodigo/dto/request/livro/NovoLivroRequest.kt
package io.github.monthalcantara.casadocodigo.dto.request.livro

import com.fasterxml.jackson.annotation.JsonProperty
import io.github.monthalcantara.casadocodigo.model.Autor
import io.github.monthalcantara.casadocodigo.model.Categoria
import io.github.monthalcantara.casadocodigo.model.Livro
import org.hibernate.validator.constraints.ISBN
import java.math.BigDecimal
import javax.validation.constraints.*

data class NovoLivroRequest(
    @field:NotBlank val titulo: String,
    @field:NotBlank @field:Size( max = 500)val resumo: String,
    @field:NotBlank val sumario: String,
    @field:NotNull @field:Min(value = 20) val preco: BigDecimal,
    @field:NotNull @field:Min(value = 100) val numeroPaginas: Int,
    @field:NotBlank @field:ISBN val isbn: String,
    @field:NotNull @JsonProperty("categoria") val idCategoria: Long,
    @field:NotNull @JsonProperty("autor") val idAutor: Long
) {
    fun toModel(autor: Autor, categoria: Categoria) = Livro(
        titulo = this.titulo,
        resumo = this.resumo,
        sumario = this.sumario,
        preco = this.preco,
        numeroPaginas = this.numeroPaginas,
        isbn = this.isbn,
        categoria = categoria,
        autor = autor
    )
}

